src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
This is code is actually relying on the global lock in current MPICH. That is the reason I do not use atomic operations for this flag. As long as one thread has to go into the critical section to touch the variable, I do not need additional protection or memory barrier which avoids instruction reordering. 
The only thing I am not sure is whether the compiler does optimization to keep reading `locked` from register.